File: dl/src/test/scala/com/intel/analytics/dllib/lib/nn/performTest/BatchNormalizationPerform.lua
--
-- Created by yansh on 16-6-22.
--

require 'nn'

forwardIterations = 10
backwardIterations = 10
forwardTimeoutMillis = 30
backwardTimeoutMillis = 120
inputNum = 100
featureDim = 512
seed = 100

torch.manualSeed(seed)
file = io.open("TorchPerform.csv", "a")
io.output(file)

bn = nn.BatchNormalization(featureDim)

for i = 1, featureDim do
    bn.weight[i] = 0.1 * i
    bn.bias[i] = 0.1 * i
end

input = torch.Tensor(inputNum, featureDim)

--warm up
for j = 1, forwardIterations do
    bn:forward(input)
end

timer = torch.Timer()
for j = 1, forwardIterations do
    bn:forward(input)
end
timeSpend = timer:time().real / forwardIterations
file:write("BatchNormalization_forward:" .. string.format("%.6f", timeSpend) * 1000 .. "\n")

grads = torch.Tensor(inputNum, featureDim)

timer = torch.Timer()
for j = 1, backwardIterations do
    bn:backward(input,grads)
end
timeSpend = timer:time().real / backwardIterations
file:write("BatchNormalization_backward:" .. string.format("%.6f", timeSpend) * 1000 .. "\n")

file:close()

File: dl/src/test/scala/com/intel/analytics/dllib/lib/nn/performTest/BatchNormalizationPerform.scala
package com.intel.analytics.dllib.lib.nn.performTest

import com.intel.analytics.dllib.lib.nn.BatchNormalization
import com.intel.analytics.dllib.lib.tensor.torch
import org.scalatest.{BeforeAndAfter, FlatSpec}
import com.intel.analytics.dllib.lib.tensor.RandomGenerator._


/**
  * Created by yao on 5/19/16.
  */
class BatchNormalizationPerform extends FlatSpec with BeforeAndAfter{
  val ForwardIterations = PerformTestData.forwardIterations
  val BackwardIterations = PerformTestData.backwardIterations
  val ForwardTimeoutMillis = 5
  val BackwardTimeoutMillis = 6
  val InputNum = 100
  val FeatureDim = 512
  val Seed = 100
  val bn = new BatchNormalization[Double](FeatureDim)

  for (i <- 1 until FeatureDim) {
    bn.weight(i) = 0.1 * i
    bn.bias(i) = 0.1 * i
  }

  RNG.setSeed(Seed)
  val input = torch.Tensor[Double](InputNum, FeatureDim).rand()

  //Warm up
  for (j <- 0 until ForwardIterations){
    bn.forward(input)
  }

  "A BatchNormalization" should " forward with a good performance" in {
    if (!TestUtils.isRun())
      cancel("Performance regression tests are forbidden")
    val timeMillis = TestUtils.testModuleForwardPerform(input, bn, ForwardIterations)
    //assert(timeMillis < ForwardTimeoutMillis)
  }

  val grads = torch.Tensor[Double](InputNum, FeatureDim).rand()

  it should(" backward with a good performance") in {
    if (!TestUtils.isRun())
      cancel("Performance regression tests are forbidden")
    val timeMillis = TestUtils.testModuleBackwardPerform(input, grads, bn, BackwardIterations)
    assert(timeMillis < BackwardTimeoutMillis)
  }
}
